### PR TITLE
Stratified Tank Optimizations

### LINKED ISSUE #5039: HPWH Stratified Tank performance is slow
When working with PR #4698, I found that the code runs pretty dang slow on my debug build.  In release it isn't too bad, but after talking with @nmerket, it seems that there is definite room for investigating some improvements.  This issue is simply a follow-on from that PR to remember to investigate that issue in the future.

### PR DESCRIPTION
**Pull request by nmerket**
- [x] Start by looking at general optimizations
  - [x] ~~fix `dt` rounding issue~~ isn't causing a problem as far as I can tell
  - [x] ~~inversion mixing at top and bottom of tank using correct temperatures~~ I tested this, and it is not an issue either because the coefficients are set to zero at those top and bottom nodes.
  - [x] `Euse` calculation using correct use mass flow rate. I added a test and a fix for this bug.
- [x] ~~See if we can get more by increasing `dt`.~~ Will deal with this later.

Fixes #5039

### CODE REVIEW

File: src/EnergyPlus/WaterThermalTanks.cc [hunk 1]
```diff
@@ -6479,64 +6483,69 @@ namespace WaterThermalTanks {
...
+				Eloss += Qloss;
+				Eneeded += Qneeded;
+				Eunmet += Qunmet;
 
 			} // NodeNum
 
-			Euse += UseMassFlowRate * Cp * (UseInletTemp - Tank.Node( Tank.UseOutletStratNode ).Temp) * dt;
+			Euse += UseMassFlowRate * ( UseInletTemp - tank_useOutletStratNode_temp );
+//? This is using the value of UseMassFlowRate from the last node
+//  If that is intended a comment would be a good idea
```
Comment (nmerket) [inline src/EnergyPlus/WaterThermalTanks.cc]: @DeadParrot I don't think that's intended. It should be the use mass flow rate of the use outlet node. I'm going to create a unit test to exercise the bug and verify that this is happening.

File: src/EnergyPlus/WaterThermalTanks.cc [hunk 2]
```diff
@@ -6479,64 +6483,69 @@ namespace WaterThermalTanks {
 					TempUp = Tank.Node( NodeNum - 1 ).Temp;
 					if ( TempUp < NodeTemp ) InvMixUp = Tank.InversionMixingRate;
 				}
+//BUG? When NodeNum==1 TempUp is 0.0   Shouldn't it be NodeTemp to void the vertical conduction? !!!!!!!!!!!!!!!!!
```
Comment (nmerket) [inline src/EnergyPlus/WaterThermalTanks.cc:6835]: @DeadParrot yes [this does look like a bug](https://github.com/NREL/EnergyPlus/commit/9efa39dbb1a6b792575f7bcd0d876cc71a0c5dee#commitcomment-16065959) to me too. I'll dig in more and probably make a unit test here too to verify before changing it. Thanks for finding it.

File: src/EnergyPlus/WaterThermalTanks.cc [hunk 3]
```diff
@@ -6368,7 +6370,9 @@ namespace WaterThermalTanks {
 		if ( Tank.InletMode == InletModeFixed ) CalcNodeMassFlows( WaterThermalTankNum, InletModeFixed );
 
 		TimeRemaining = SecInTimeStep;
-		while ( TimeRemaining > 0.0 ) {
+		Real64 & tank_node1_newTemp( Tank.Node( 1 ).NewTemp );
+		Real64 const & tank_useOutletStratNode_temp( Tank.Node( Tank.UseOutletStratNode ).Temp );
+		while ( TimeRemaining > 0.0 ) { //? Might be good to use a check that is tolerant to some epsilon * dt to avoid an extra pass due to roundoff when subtracting dt (if dt is ever not a value with an exact floating point representation)
```
Comment (nmerket) [inline src/EnergyPlus/WaterThermalTanks.cc:6724]: @DeadParrot, I'll make sure to [address this](https://github.com/NREL/EnergyPlus/commit/9efa39dbb1a6b792575f7bcd0d876cc71a0c5dee#commitcomment-16066008) too. Thanks.

**Comment by mbadams5** [comment]
Not to be a naysayer (and I know this is just a first pass at optimizations) but I would be surprised if these changes affect performance in Release builds in any significant way. They probably help Debug builds but we shouldn't try to optimize around Debug builds. The compiler is usually pretty smart about optimizations in Release builds and if it is not then that is where you focus on changes.

I will say though, if these kinds of changes DO have an impact on Release performance then these are easy changes to address in other parts of the code as well.

I just know from experience that when I try to outsmart the compiler about optimizations, I sometimes lose. Meaning the compiler was already doing the types of optimizations I attempted to do by hand.

**Comment by nmerket** [comment]
Well, that's disappointing. I just compiled release builds of before and after and ran them with a model that calls `CalcWaterThermalTankStratified` a lot. I was hoping to prove you wrong, but here's what I found:

### Before

```
real    2m59.320s
user    2m56.783s
sys 0m1.883s
```

### After

```
real    2m56.993s
user    2m54.636s
sys 0m1.901s
```

So, yeah, not really any difference.

**Comment by DeadParrot** [comment]
@mbadams5 These changes gave ~20% speedup using Intel C++ on Windows for the (annualized) run time of the test case we are using. Some of the changes are more useful than others, of course. I may not separately test each change but I do test and profile (release builds) when I'm tuning. I did not find any huge gains to be had and I left it to @nmerket to decide which of the changes to keep -- some may not be worth it if people find the code confusing.

I'll try the annualized run on GCC to see if the changes are not effective there.

**Comment by nmerket** [comment]
I was running it on an annual run of `testfiles/WaterHeaterHeatPumpWrappedCondenser.idf`. I used clang on the mac to do the release build. I'm building it to run through the Xcode profiler now...

**Comment by mbadams5** [comment]
@DeadParrot You saw a 20% speedup on a full annual run? Especially when you said 

> I did not find any huge gains to be had...

I mean a 20% speedup on an annual run is huge...

**Comment by DeadParrot** [comment]
@mbadams5 Yes, 20% speedup on the annual run. I don't consider 20% huge but it certainly can be important.

**Comment by nmerket** [comment]
Yeah, I just ran before and after through the profiler on my machine with release builds and I'm not seeing any speedup. I am seeing that 70% of the computation time for each is spent on `CalcWaterThermalTankStratified`, though.

**Comment by DeadParrot** [comment]
So I'm only getting ~5% speedup with GCC 5.3.1 but I'm seeing ~25% with Visual C++ 2015 so it definitely is varying, which isn't unusual. At least in no case is it slower ;-).

**Comment by DeadParrot** [comment]
And I only get ~3.5% speedup on Clang 3.7 on Linux. So probably the Linux compiler optimizers are better at "auto-hoisting".

Big picture: none of this really addresses the big performance bottleneck of this algorithm. If increasing dt isn't sufficient then a deeper look at the algorithm and data structures will be needed.

**Comment by nmerket** [comment]
Well the 25% speedup on Windows makes it worth these changes to me. I agree that there are bigger issues to deal with here, I just wanted to make sure we got the straightforward optimizations out of the way before plowing forward with changes that could change the simulation results.

**Comment by nrel-bot-3** [comment]
@nmerket @lgentile it has been 14 days since this pull request was last updated.

**Comment by mjwitte** [comment]
@nmerket Is this intended for v8.5 or later?

**Comment by nmerket** [comment]
I think there is more we'd like to do here before merging, although this certainly could go into 8.5 and would provide the modest speed ups we discussed sooner rather than later.

**Comment by mjwitte** [comment]
OK. I'm going to mark it 8.5 for now, and you and @Myoldmopar can decide whether to merge what's here or move it to 8.6.

**Comment by Myoldmopar** [comment]
With the bug fixes intermingled with the performance changes, I'm inclined to let it hold off until after the release.  Maybe with some more time spent on the bug fixes and performance, the code readability could be improved too.

**Comment by nrel-bot-3** [comment]
@nmerket @lgentile it has been 14 days since this pull request was last updated.

**Comment by nrel-bot-3** [comment]
@nmerket @lgentile it has been 14 days since this pull request was last updated.

**Comment by nrel-bot-3** [comment]
@nmerket @lgentile it has been 14 days since this pull request was last updated.

**Comment by nrel-bot-2** [comment]
@nmerket @lgentile it has been 14 days since this pull request was last updated.

**Comment by Myoldmopar** [comment]
@nmerket What else do you want to do on this one?

**Comment by Myoldmopar** [comment]
@nmerket I went ahead and merged this up and pushed to let CI have another crack at it.  Were you planning more work here?  There's no new testing, although that's not necessary if existing unit tests are already covering the changed code (the unit test would fail if these changes caused diffs...).

**Comment by nmerket** [comment]
I totally forgot about this. Go ahead and merge it. I would like to do more to speed up the stratified tanks, but that might be down the road a little bit. I can do that on a separate pull request.

**Comment by Myoldmopar** [comment]
Is there already a unit test that covers this section of the code, @nmerket?

**Comment by Myoldmopar** [comment]
Nevermind, I see it.  It is somewhat covered: http://energyplus.s3-website-us-east-1.amazonaws.com/coverage/2016-02-12-EnergyPlus-9efa39dbb1-x86_64-gcc-4.8UTCD/EnergyPlus/WaterThermalTanks.cc.gcov.html

In order to wrap this up, could you improve the unit test coverage of this function to make sure it is good to go?

**Comment by nmerket** [comment]
I will see what I can do. This can be done after the idd freeze, right?

**Comment by Myoldmopar** [comment]
Yeah, for sure.

**Comment by nmerket** [comment]
@Myoldmopar this should be ready to go. Merge away.

**Comment by nrel-bot-3** [comment]
@nmerket @lgentile it has been 14 days since this pull request was last updated.

**Comment by nmerket** [comment]
@Myoldmopar any idea when this will merge? @shorowit is dealing with some really slow HPWH simulations and was expecting this to be in 8.6.

**Comment by nmerket** [comment]
ping @Myoldmopar

**Pull request merged**